### PR TITLE
Improvements to `PhasingViaCostFunction` infrastructure

### PR DESCRIPTION
**Pull request by tanujkhattar**
Fixes https://github.com/quantumlib/Qualtran/issues/654

### CODE REVIEW

File: qualtran/bloqs/rotations/phasing_via_cost_function_test.py [hunk 1]
```diff
@@ -75,7 +77,7 @@ def build_composite_bloq(self, bb: 'BloqBuilder', **soqs: 'SoquetT') -> Dict[str
 def test_hamming_weight_phasing_using_phase_via_cost_function_quick():
     n = 2
     exponent = 1.20345
-    eps = 5e-4
+    eps = 1e-3
```
Comment (fdmalone) [inline qualtran/bloqs/rotations/phasing_via_cost_function_test.py:80]: Why did this change?

File: qualtran/bloqs/rotations/phase_gradient.py [hunk 2]
```diff
@@ -149,26 +147,46 @@ class AddIntoPhaseGrad(GateWithRegisters, cirq.ArithmeticGate):
         [Compilation of Fault-Tolerant Quantum Heuristics for Combinatorial Optimization]
         (https://arxiv.org/abs/2007.07391), Appendix A: Addition for controlled rotations
     """
-    inp_bitsize: int
+    x_bitsize: int
     phase_bitsize: int
+    right_shift: int = 0
```
Comment (fdmalone) [inline qualtran/bloqs/rotations/phase_gradient.py:152]: update docstring.

File: qualtran/bloqs/rotations/phase_gradient.py [hunk 3]
```diff
@@ -135,10 +135,8 @@ class AddIntoPhaseGrad(GateWithRegisters, cirq.ArithmeticGate):
         U|x\rangle|\text{phase\_grad}\rangle = |x\rangle|\text{phase\_grad} + x\rangle
     $$
 
-    TODO(#654): Check whether phase_bitsize >= inp_bitsize needs to be enforced.
```
Comment (fdmalone) [inline qualtran/bloqs/rotations/phase_gradient.py:138]: Should probably capture the conclusion of this todo in the args somewhere.

File: qualtran/bloqs/rotations/phase_gradient.py [hunk 4]
```diff
@@ -149,26 +147,46 @@ class AddIntoPhaseGrad(GateWithRegisters, cirq.ArithmeticGate):
...
     def with_registers(self, *new_registers: Union[int, Sequence[int]]):
         raise NotImplementedError("not needed.")
 
-    def apply(self, x, phase_grad) -> Union[int, Iterable[int]]:
-        return x, phase_grad + x
+    @cached_method
+    def scaled_val(self, x: int) -> int:
+        """Computes `phase_grad + x` using fixed point arithmetic."""
+        x_width = self.x_bitsize + self.right_shift
+        x_fxp = _fxp(x / 2**x_width, x_width).like(_fxp(0, self.phase_bitsize)).astype(float)
```
Comment (fdmalone) [inline qualtran/bloqs/rotations/phase_gradient.py:180]: am i reading this line right:

1. x / 2**x_width of width x_width 
2. Convert to a new fxp of width phase_bitsize
3. convert to float.

Why is step 2 required?

File: qualtran/bloqs/rotations/phase_gradient.py [hunk 5]
```diff
@@ -149,26 +147,46 @@ class AddIntoPhaseGrad(GateWithRegisters, cirq.ArithmeticGate):
...
+        return int(x_fxp.astype(float) * 2**self.phase_bitsize)
+
+    def apply(self, x: int, phase_grad: int) -> Union[int, Iterable[int]]:
+        out = self.on_classical_vals(x=x, phase_grad=phase_grad)
+        return out['x'], out['phase_grad']
 
-    def on_classical_vals(self, x, phase_grad) -> Dict[str, 'ClassicalValT']:
-        return {'x': x, 'phase_grad': (phase_grad + x) % (2**self.phase_bitsize)}
+    def on_classical_vals(self, x: int, phase_grad: int) -> Dict[str, 'ClassicalValT']:
```
Comment (fdmalone) [inline qualtran/bloqs/rotations/phase_gradient.py:187]: I'm a bit confused, so on input x should be the the integer representation of what we want our fixed point real number to be. What values can x take?

File: qualtran/bloqs/rotations/phase_gradient.py [hunk 6]
```diff
@@ -179,6 +197,33 @@ def _t_complexity_(self) -> 'TComplexity':
         return n * toffoli.t_complexity()
 
 
+def _fxp(x: float, out: int):
+    assert 0 <= x < 1
```
Comment (fdmalone) [inline qualtran/bloqs/rotations/phase_gradient.py:201]: maybe add a little context here so we don't have to google the Fxp kwargs.

File: qualtran/bloqs/rotations/phase_gradient.py [hunk 7]
```diff
@@ -237,12 +290,11 @@ def scaled_val(self, x: int) -> int:
         # However, `x` should be interpreted as per the fixed point specification given in self.inp_dtype.
```
Comment (fdmalone) [inline qualtran/bloqs/rotations/phase_gradient.py:290]: inp_dtype is no more.

File: qualtran/bloqs/rotations/phase_gradient.py [hunk 8]
```diff
@@ -224,11 +269,19 @@ def with_registers(self, *new_registers: Union[int, Sequence[int]]):
...
+    @cached_property
+    def inp_dtype(self) -> QFxp:
+        return QFxp(self.x_bitsize, self.x_bitsize, signed=False)
+
     @cached_property
     def gamma_dtype(self) -> QFxp:
         n_int = Fxp(abs(self.gamma), signed=False).n_int
         return QFxp(n_int + self.gamma_bitsize, self.gamma_bitsize, signed=False)
 
+    @cached_property
```
Comment (fdmalone) [inline qualtran/bloqs/rotations/phase_gradient.py:281]: Why do we take abs(gamma) and not just keep it signed?

File: qualtran/bloqs/rotations/phasing_via_cost_function.py [hunk 9]
```diff
@@ -154,30 +158,20 @@ def b_phase(self) -> int:
...
-        # Using Equation 35 from https://arxiv.org/abs/2007.07391
-        eq_35 = self.b_phase + int(np.ceil(np.log2(self.b_phase)))
-        # TODO(#654): Eq A7 will result in a bigger gradient bitsize but blows up the cost
-        #   for doing phase gradient based cost computations significantly (which leads to
-        #   the cost using PhaseOracleZPow to be cheaper). Also, we don't yet have a test that
-        #   fails for Eq 35. The only concern is that value of `eq_35` can be smaller than
-        #   `cost_reg.bitsize`.
-        assert eq_a7 >= eq_35
-        return eq_35
+        return int(np.ceil(np.log2((self.gamma_bitsize + 2) * np.pi / self.eps)))
```
Comment (fdmalone) [inline qualtran/bloqs/rotations/phasing_via_cost_function.py:161]: What was the resolution here? A7 is correct?

File: qualtran/bloqs/rotations/phasing_via_cost_function.py [hunk 10]
```diff
@@ -154,30 +158,20 @@ def b_phase(self) -> int:
...
-        # Note: Paragraph b/w equation 34 & 35 of https://arxiv.org/abs/2007.07391 gives
-        # `gamma_bitsize` to be `log(gamma) + b_{phase} + O(1)`. However, this is incorrect, and
-        # we have tests that fail if you do `return self.b_phase`.
-        # The correct `gamma_bitsize` can be obtained using Equation D7 and is given below.
-        d_B = self.cost_dtype.bitsize
-        return d_B + int(np.ceil(np.log2(d_B / self.eps)))  # Using Equation D7
+        # Using `gamma_bitsize = log(gamma) + b_{phase} + O(1)` defined b/w equation 34 & 35
+        # of https://arxiv.org/abs/2007.07391. Note that the `log(gamma)` here means ignoring
+        # leading digits of `gamma` which are 0
+        return self.b_phase
```
Comment (fdmalone) [inline qualtran/bloqs/rotations/phasing_via_cost_function.py:168]: We can get away with the + O(1) here? Should this be determined through some simulation?

**Review by fdmalone — CHANGES_REQUESTED** [review]
I just left several questions. Maybe worth using this time to add the autodoc examples to capture the nice docstrings.

File: qualtran/bloqs/rotations/hamming_weight_phasing.py [hunk 11]
```diff
@@ -168,7 +175,10 @@ def build_composite_bloq(
         x, junk, out = bb.add(HammingWeightCompute(self.bitsize), x=x)
         out, phase_grad = bb.add(
             AddScaledValIntoPhaseReg(
-                self.phase_oracle.cost_dtype, self.b_grad, self.exponent / 2, self.gamma_bitsize
+                self.phase_oracle.cost_dtype.bitsize,
+                self.b_grad,
+                (self.exponent / 2) * (2 ** self.bitsize.bit_length()),
+                self.gamma_bitsize,
```
Comment (mpharrigan) [inline qualtran/bloqs/rotations/hamming_weight_phasing.py:181]: keyword arguments would not be amiss here

**Comment by mpharrigan** [comment]
+1 to @fdmalone 's comments about recording the result of this investigation in the codebase

File: qualtran/bloqs/rotations/phasing_via_cost_function_test.py (hunk 1 above)
Comment (tanujkhattar) [reply qualtran/bloqs/rotations/phasing_via_cost_function_test.py:87]: to make the test run quickly

File: qualtran/bloqs/rotations/phase_gradient.py (hunk 2 above)
Comment (tanujkhattar) [reply qualtran/bloqs/rotations/phase_gradient.py:156]: done

File: qualtran/bloqs/rotations/phase_gradient.py (hunk 3 above)
Comment (tanujkhattar) [reply qualtran/bloqs/rotations/phase_gradient.py:138]: See the newly added and very detailed `qualtran/bloqs/rotations/quantum_variable_rotation.ipynb` notebook.

File: qualtran/bloqs/rotations/phase_gradient.py (hunk 5 above)
Comment (tanujkhattar) [reply qualtran/bloqs/rotations/phase_gradient.py:191]: Ideally input `x` would be an `Fxp` object instead of `int` but we don't have support for quantum data types in classical simulation yet. I've added more details in the newly added notebook `qualtran/bloqs/rotations/quantum_variable_rotation.ipynb`

File: qualtran/bloqs/rotations/phase_gradient.py (hunk 6 above)
Comment (tanujkhattar) [reply qualtran/bloqs/rotations/phase_gradient.py:215]: done

File: qualtran/bloqs/rotations/phase_gradient.py (hunk 8 above)
Comment (tanujkhattar) [reply qualtran/bloqs/rotations/phase_gradient.py]: since `gamma` is a classically known constant, we can decide at the time of circuit construction whether we wish to add or subtract the scaled value into the target register. Thus, we don't need to waste one bit of gamma to store the sign information. This is why `AddIntoPhaseGrad` accepts a `sign: int` which is essentially the `sign(gamma)`

File: qualtran/bloqs/rotations/phasing_via_cost_function.py (hunk 9 above)
Comment (tanujkhattar) [reply qualtran/bloqs/rotations/phasing_via_cost_function.py]: See the newly added notebook `qualtran/bloqs/rotations/quantum_variable_rotation.ipynb`. 

Tl;Dr is that both equations were correct and equivalent given the assumptions in the paper. In the general case, A7 is more accurate.

File: qualtran/bloqs/rotations/phasing_via_cost_function.py (hunk 10 above)
Comment (tanujkhattar) [reply qualtran/bloqs/rotations/phasing_via_cost_function.py]: The + O(1) is due log(2pi); updated the implementation and added lots of details in the new notebook.

File: qualtran/bloqs/rotations/hamming_weight_phasing.py (hunk 11 above)
Comment (tanujkhattar) [reply qualtran/bloqs/rotations/hamming_weight_phasing.py]: added

**Comment by tanujkhattar** [comment]
See qualtran/bloqs/rotations/quantum_variable_rotation.ipynb for a very detailed pedagogical review (which I think I've done better than the linked references, and also explained a more general case) of all the details behind QVR via phase gradients. 

I've also added LOTS of more docstrings and auto generated notebooks. Please take another look. 

Since the size of the PR has grown significantly and there are a lot of nuanced details; I'd like to get this merged soon to avoid merge conflicts with anything else. I'll present the detailed review in the Qualtran sync tomorrow as well.

File: qualtran/bloqs/rotations/quantum_variable_rotation.ipynb [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+    "\n",
+    "A number of subtleties arise as part of this procedure and we describe them below one by one.\n",
+    "\n",
+    "- **Adding a scaled value into phase gradient register** Instead of computing $\\gamma x$ an intermediate register, we perform the multiplication via repeated additions into the phase gradient register, as described in [2]. This requires us to represent $\\gamma$ as a fixed point fraction with bitsize $\\gamma_\\text{bitsize}$. This procedure introduces two sources of errors:\n",
+    "    - **Errors due to fixed point representation of $\\gamma$** - Note that adding any fixed point number of the form $a.b$ to the phase gradient register is equivalent to adding $0.b$ since $e^{2\\pi i a} = 1$ for every integer $a$. Let $\\tilde{\\gamma} = a.b$ and $x = p.q$ be fixed point decimal representations of $\\gamma$ and $x$. We can write the product $\\gamma x$ as\n",
+    "    $$\n",
+    "          \\tilde{\\gamma} x = (\\sum_{i=0}^{\\gamma_\\text{n\\_int}} a_{i} * 2^{i} + \\sum_{i=1}^{\\gamma_\\text{n\\_frac}} \\frac{b_i}{2^i}) (\\sum_{j=0}^{x_\\text{n\\_int}} p_{j} * 2^{j} + \\sum_{j=1}^{x_\\text{n\\_frac}} \\frac{q_{j}}{2^{j}})\n",
```
Comment (fdmalone) [inline qualtran/bloqs/rotations/quantum_variable_rotation.ipynb:220]: nit: the *'s here should probably be \times to render a bit nicer.

File: qualtran/bloqs/rotations/quantum_variable_rotation.ipynb [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+    "                                &\\approxeq \\log_2{\\frac{1}{\\epsilon}} + O(1)\n",
+    "    \\end{aligned}$$\n",
+    "\n",
+    "    - **Errors due to truncation of digits of $|x\\rangle$ during multiplication via repeated addition** - Let $b_\\text{grad}$ be the size of the phase gradient register. When adding right shifted copies of state $x$ to the phase gradient register, we incur an error every time the fractional part of the right shifted state to be added needs to be truncated to $b_\\text{grad}$ digits. For each such addition the error is upper bounded by $\\frac{2\\pi}{2^{b_\\text{grad}}}$, because we omit adding bits that would correspond to phase shifts of $\\frac{2\\pi}{2^{b_\\text{grad}+1}}$, $\\frac{2\\pi}{2^{b_\\text{grad}+2}}$, and so forth. The number of such additions can be upper bounded by $\\frac{(\\gamma_\\text{n\\_frac} + 2)}{2}$ using techniques from [2]. When $b_\\text{grad} >= x_\\text{bitsize}$, the first $x_\\text{n\\_int}$ additions do not contribute to any phase error and thus the number of error causing additions can be upper bounded by $\\frac{(b_\\text{phase} + 2)}{2}$. In order to keep the error less then $\\epsilon$, we get\n",
```
Comment (fdmalone) [inline qualtran/bloqs/rotations/quantum_variable_rotation.ipynb:229]: nit: >= -> \ge

**Review by fdmalone — APPROVED** [review]

File: qualtran/bloqs/rotations/quantum_variable_rotation.py [hunk 14]
```diff
@@ -0,0 +1,339 @@
...
+)
+
+
+@attrs.frozen
+class QvrPhaseGradient(QvrInterface):
+    r"""QVR oracle that applies a rotation via addition into the phase gradient register.
+
+    A $b_\text{grad}$-bit phase gradient state $|\phi\rangle_{b_\text{grad}}$ can be written as
+
+    <!-- can be thought of as a superposition of states $|\frac{0}{2^{b_\text{grad}}}\rangle, |\frac{2}{2^{b_\text{grad}}}\rangle, |\frac{2}{2^{b_\text{grad}}}\rangle, ... |k/$ where each integer $0\leq k \lt 2^{b_\text{grad}}$ represents a $b_\text{grad}$-bit fixed point number $\frac{k}{2^{b_\text{grad}}}$ -->
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply qualtran/bloqs/rotations/quantum_variable_rotation.py:176]: because of the line length? yeah, curious.

What's with the html-style comments? are these omitted from the notebook?

[the remark this answers is not in the document]
Comment (mpharrigan) [reply qualtran/bloqs/rotations/quantum_variable_rotation.py:176]: all of the "C" lint  check s are disabled including `line-too-long` since the formatter usually handles it.

[the remark this answers is not in the document]
Comment (mpharrigan) [reply qualtran/bloqs/rotations/quantum_variable_rotation.py:176]: but please wrap these docstrings. Your IDE should be able to do most of it for you

[the remark this answers is not in the document]
Comment (fdmalone) [reply qualtran/bloqs/rotations/quantum_variable_rotation.py:176]: black can't handle docstrings right

[the remark this answers is not in the document]
Comment (tanujkhattar) [reply qualtran/bloqs/rotations/quantum_variable_rotation.py]: wrapped the docstrings, removed html comments (leftover by mistake)

File: qualtran/bloqs/rotations/quantum_variable_rotation.ipynb (hunk 13 above)
Comment (tanujkhattar) [reply qualtran/bloqs/rotations/quantum_variable_rotation.ipynb]: done

**Comment by tanujkhattar** [comment]
I've updated the docstrings and added some more tests to `quantum_variable_rotation_test.py`. I'm unable to prove the right precision for the case when `b_grad < x_bitsize` but I believe the precision should depend only upon `log(1/eps)` and not scale with size of the cost register. I've update the docstrings to include this information, added some worst case tests where `b_grad < x_bitsize`.  I'll merge this PR now and we can revisit later if we discover this assumption to be incorrect.
